File: hbase-client/src/main/java/org/apache/hadoop/hbase/client/AsyncTableResultScanner.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.hadoop.hbase.client;

import static org.apache.hadoop.hbase.client.ConnectionUtils.calcEstimatedSize;
import io.opentelemetry.api.trace.Span;
import java.io.IOException;
import java.io.InterruptedIOException;
import java.util.ArrayDeque;
import java.util.Queue;
import org.apache.hadoop.hbase.TableName;
import org.apache.hadoop.hbase.client.metrics.ScanMetrics;
import org.apache.hadoop.hbase.util.FutureUtils;
import org.apache.yetus.audience.InterfaceAudience;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * The {@link ResultScanner} implementation for {@link RawAsyncTableImpl}. It will fetch data
 * automatically in background and cache it in memory. Typically, the {@link #maxCacheSize} will be
 * {@code 2 * scan.getMaxResultSize()}.
 */
@InterfaceAudience.Private
class AsyncTableResultScanner implements ResultScanner, AdvancedScanResultConsumer {

  private static final Logger LOG = LoggerFactory.getLogger(AsyncTableResultScanner.class);

  private final TableName tableName;

  private final long maxCacheSize;

  private final Scan scan;

  private final Queue<Result> queue = new ArrayDeque<>();

  private ScanMetrics scanMetrics;

  private long cacheSize;

  private boolean closed = false;

  private Throwable error;

  private ScanResumer resumer;

  // Used to pass the span instance to the `AsyncTableImpl` from its underlying `rawAsyncTable`.
  private Span span = null;

  public AsyncTableResultScanner(TableName tableName, Scan scan, long maxCacheSize) {
    this.tableName = tableName;
    this.maxCacheSize = maxCacheSize;
    this.scan = scan;
  }

  private void addToCache(Result result) {
    queue.add(result);
    cacheSize += calcEstimatedSize(result);
  }

  private void stopPrefetch(ScanController controller) {
    if (LOG.isDebugEnabled()) {
      LOG.debug("{} stop prefetching when scanning {} as the cache size {}" +
        " is greater than the maxCacheSize {}",
        String.format("0x%x", System.identityHashCode(this)), tableName, cacheSize,
        maxCacheSize);
    }
    resumer = controller.suspend();
  }

  Span getSpan() {
    return span;
  }

  void setSpan(final Span span) {
    this.span = span;
  }

  @Override
  public synchronized void onNext(Result[] results, ScanController controller) {
    assert results.length > 0;
    if (closed) {
      controller.terminate();
      return;
    }
    for (Result result : results) {
      addToCache(result);
    }
    notifyAll();
    if (cacheSize >= maxCacheSize) {
      stopPrefetch(controller);
    }
  }

  @Override
  public synchronized void onHeartbeat(ScanController controller) {
    if (closed) {
      controller.terminate();
      return;
    }
    if (scan.isNeedCursorResult()) {
      controller.cursor().ifPresent(c -> queue.add(Result.createCursorResult(c)));
    }
  }

  @Override
  public synchronized void onError(Throwable error) {
    this.error = error;
    notifyAll();
  }

  @Override
  public synchronized void onComplete() {
    closed = true;
    notifyAll();
  }

  @Override
  public void onScanMetricsCreated(ScanMetrics scanMetrics) {
    this.scanMetrics = scanMetrics;
  }

  private void resumePrefetch() {
    if (LOG.isDebugEnabled()) {
      LOG.debug(String.format("0x%x", System.identityHashCode(this)) + " resume prefetching");
    }
    resumer.resume();
    resumer = null;
  }

  @Override
  public synchronized Result next() throws IOException {
    while (queue.isEmpty()) {
      if (closed) {
        return null;
      }
      if (error != null) {
        throw FutureUtils.rethrow(error);
      }
      try {
        wait();
      } catch (InterruptedException e) {
        throw new InterruptedIOException();
      }
    }
    Result result = queue.poll();
    if (!result.isCursor()) {
      cacheSize -= calcEstimatedSize(result);
      if (resumer != null && cacheSize <= maxCacheSize / 2) {
        resumePrefetch();
      }
    }
    return result;
  }

  @Override
  public synchronized void close() {
    closed = true;
    queue.clear();
    cacheSize = 0;
    if (resumer != null) {
      resumePrefetch();
    }
    notifyAll();
  }

  @Override
  public boolean renewLease() {
    // we will do prefetching in the background and if there is no space we will just suspend the
    // scanner. The renew lease operation will be handled in the background.
    return false;
  }

  // used in tests to test whether the scanner has been suspended
  synchronized boolean isSuspended() {
    return resumer != null;
  }

  @Override
  public ScanMetrics getScanMetrics() {
    return scanMetrics;
  }

  int getCacheSize() {
    return queue.size();
  }
}

File: src/hbase/client/async-table-result-scanner.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include "hbase/client/async-table-result-scanner.h"

#include <vector>

namespace hbase {
AsyncTableResultScanner::AsyncTableResultScanner(int64_t max_cache_size)
    : max_cache_size_(max_cache_size) {
  closed_ = false;
  cache_size_ = 0;
}

AsyncTableResultScanner::~AsyncTableResultScanner() { Close(); }

void AsyncTableResultScanner::Close() {
  std::unique_lock<std::mutex> mlock(mutex_);
  closed_ = true;
  while (!queue_.empty()) {
    queue_.pop();
  }
  cache_size_ = 0;
  if (resumer_ != nullptr) {
    resumer_->Resume();
  }
  cond_.notify_all();
}

std::shared_ptr<Result> AsyncTableResultScanner::Next() {
  VLOG(5) << "AsyncTableResultScanner: Next()";

  std::shared_ptr<Result> result = nullptr;
  std::shared_ptr<ScanResumer> local_resumer = nullptr;
  {
    std::unique_lock<std::mutex> mlock(mutex_);
    while (queue_.empty()) {
      if (closed_) {
        return nullptr;
      }
      if (error_) {
        throw error_;
      }
      cond_.wait(mlock);
    }
    result = queue_.front();
    queue_.pop();

    cache_size_ -= EstimatedSizeWithSharedPtr(result);
    if (resumer_ != nullptr && cache_size_ <= max_cache_size_ / 2) {
      VLOG(1) << std::this_thread::get_id() << " resume scan prefetching";
      local_resumer = resumer_;
      resumer_ = nullptr;
    }
  }

  // Need to call ScanResumer::Resume() outside of the scope of the mutex. The reason is that
  // folly/wangle event loop might end up running the attached logic(.then()) at the Scan RPC
  // in the same event thread before returning from the previous call. This seems like the
  // wrong thing to do(™), but we cannot fix that now. Since the call back can end up calling
  // this::OnNext(), we should unlock the mutex.
  if (local_resumer != nullptr) {
    local_resumer->Resume();
  }
  return result;
}

void AsyncTableResultScanner::AddToCache(const std::vector<std::shared_ptr<Result>> &results) {
  VLOG(5) << "AsyncTableResultScanner: AddToCache()";
  for (const auto r : results) {
    queue_.push(r);
    cache_size_ += EstimatedSizeWithSharedPtr(r);
  }
}

template <typename T>
inline size_t AsyncTableResultScanner::EstimatedSizeWithSharedPtr(std::shared_ptr<T> t) {
  return t->EstimatedSize() + sizeof(std::shared_ptr<T>);
}

void AsyncTableResultScanner::OnNext(const std::vector<std::shared_ptr<Result>> &results,
                                     std::shared_ptr<ScanController> controller) {
  VLOG(5) << "AsyncTableResultScanner: OnNext()";
  {
    std::unique_lock<std::mutex> mlock(mutex_);
    if (closed_) {
      controller->Terminate();
      return;
    }
    AddToCache(results);

    if (cache_size_ >= max_cache_size_) {
      StopPrefetch(controller);
    }
  }
  cond_.notify_all();
}

void AsyncTableResultScanner::StopPrefetch(std::shared_ptr<ScanController> controller) {
  VLOG(1) << std::this_thread::get_id()
          << ": stop prefetching when scanning as the cache size " +
                 folly::to<std::string>(cache_size_) + " is greater than the max_cache_size " +
                 folly::to<std::string>(max_cache_size_);

  resumer_ = controller->Suspend();
  num_prefetch_stopped_++;
}

/**
 * Indicate that there is an heartbeat message but we have not cumulated enough cells to call
 * onNext.
 * <p>
 * This method give you a chance to terminate a slow scan operation.
 * @param controller used to suspend or terminate the scan. Notice that the {@code controller}
 *          instance is only valid within the scope of onHeartbeat method. You can only call its
 *          method in onHeartbeat, do NOT store it and call it later outside onHeartbeat.
 */
void AsyncTableResultScanner::OnHeartbeat(std::shared_ptr<ScanController> controller) {
  std::unique_lock<std::mutex> mlock(mutex_);
  if (closed_) {
    controller->Terminate();
  }
}

/**
 * Indicate that we hit an unrecoverable error and the scan operation is terminated.
 * <p>
 * We will not call {@link #onComplete()} after calling {@link #onError(Throwable)}.
 */
void AsyncTableResultScanner::OnError(const folly::exception_wrapper &error) {
  LOG(WARNING) << "Scanner received error" << error.what();
  std::unique_lock<std::mutex> mlock(mutex_);
  error_ = error;
  cond_.notify_all();
}

/**
 * Indicate that the scan operation is completed normally.
 */
void AsyncTableResultScanner::OnComplete() {
  std::unique_lock<std::mutex> mlock(mutex_);
  closed_ = true;
  cond_.notify_all();
}
}  // namespace hbase
